proposals/time-stamp-counter-support.md

```diff
@@ -0,0 +1,274 @@
+
+### The creation of time._tsc
+
+The time module is the most logical place for most of this functionality to
+live, since they are time stamp counters. In this module, different
+architectures will be unified into a single portable API consisting of the
+following:
+
+```mojo
+@value
+struct TscFrequency:
+    """
+    Contains the TSC frequency with architecture-defined units. 
+    """
+    var _tsc_frequency: UInt64
+
+    fn get_frequency_hz(borrowed self) -> UInt64:
+        """
+        Get the frequency in Hertz. 
+        """
+        pass
+
+@value
+struct TscTimestamp:
+    """
+    Encapsulates the value of a TSC read. Units are architecture dependent.
+
+    Implementation detail: Can only reliably hold up to 2^64-1 nanoseconds or ~580 years on most architectures. 
+    """
+    var _tsc_value: UInt64
+
+    fn time_since_ns(borrowed self, borrowed other: Self, borrowed freq: TscFrequency) -> Int64:
+        """
+        Returns the time since the previous timestamp in nanoseconds using the provided frequency reference.
+        """
+        ...
+
+    fn get_raw(borrowed self) -> UInt64:
+        """
+        Access the raw, architecture dependent value of the TSC. 
+        """
+        return self._tsc_value
+
+struct Tsc:
+    """
+    Time Stamp Counter (TSC) container to hold the expensive initialization of 
+    the TSC API, and also allows for converting tsc timestamps into values 
+    usable by other APIs. Treats start_os_timestamp and start_tsc_timestamp as
+    referring to the same instant in time.
+
+    WARNING: There are hazards to the use of this API, please read the module
+    documentation, specifically the "Limitations, Hazards and Cautions" section.
+
+    freq: The frequency of the TSC, and where most of the expensive 
+    initialization comes from.
+
+    start_os_timestamp: Number of nanoseconds since an undefined point in time 
+    (typically the platform epoch). Used to aid in converting TscTimestamp 
+    instances to timestamps
+
+    start_tsc_timestamp: A tsc timestamp taken right after the timestamp 
+    syscall call returned. 
+
+    tsc_is_reliable: Whether the TSC is usable as a time source. If false, a monotonic but fast to query OS clock is used.
+    """
+
+    var freq: TscFrequency
+    var start_os_timestamp: UInt64
+    var start_tsc_timestamp: TscTimestamp
+    var tsc_is_reliable: Bool
+
+    fn __init__(inout self):
+        pass
+
+    fn get_frequency_hz(borrowed self) -> UInt64:
+        """
+        Get a TscFrequency object which can be used to get the time delta between
+        two TscTimestamp objects.
+        """
+        return self.freq.get_frequency_hz()
+
+    fn get_tsc_timestamp(borrowed self) -> TscTimestamp:
+        """
+        Read from the TSC (or fallback clock). 
+        """
+        pass
+
+    fn now(borrowed self) -> Time:
+        """
+        To be added once the standard library decides how to handle time.
```

Comment:
**Question** Have you given much thought on how this work ties into a `chrono` module and what that may look like with regards to this lower-level pre-requisite work?

Reply:
My plan was to store a unix epoch timestamp in nanoseconds (ideally in the TAI timezone) to use to convert TSC timestamps to real time. There's not a lot of value vs calling `clock_gettime()` on Linux there unless you already have TSC readings you want to convert to real time since VDSO makes `clock_gettime()` cheap, but I'd need to check the MacOS and Windows behaviors to see if this would be useful for timekeeping. 

Due to the number of hardware/firmware/hypervisor bugs that TSCs seem to have, I don't think backing a `chrono` module with them by default is a good idea, but allowing a conversion to a type from `chrono` if the user asks for it should be fine.